File: src/test/resources/jython/motionCorrection.py
# CLIJ example ImageJ Ops jython: minimum.py
#
# This macro shows how apply a minimum filter in the GPU.
#
# Author: Robert Haase, rhaase@mpi-cbg.de
# Author: Deborah Schmidt, frauzufall@mpi-cbg.de
# May 2019
# ---------------------------------------------

#@ IOService io
#@ UIService ui
#@ OpService ops

from org.scijava.table import DefaultGenericTable
import os
import inspect
def getResource(file):
	return os.path.dirname(os.path.abspath(inspect.getsourcefile(lambda:0))) + "/" + file

# define a threshold to differentiate object and background
threshold = 50

input = io.open(getResource("../samples/motion_correction_Drosophila_DSmanila1.tif"))
ui.show("input", input)
input32 = ops.convert().float32(input)
# push image to GPU
inputGPU = ops.run("CLIJ_push", input)

formerX = 0
formerY = 0

mass = DefaultGenericTable()

# process all slices; only the first stays where it is
for z in range(0, input.dimension(2)):

	slice = ops.run("CLIJ_copySlice", inputGPU, z)

	# determine center of mass
	binary = ops.run("CLIJ_threshold", slice, threshold)
	ops.run("CLIJ_centerOfMass", mass, binary)
	x = mass.get("MassX", mass.getRowCount()-1)
	y = mass.get("MassY", mass.getRowCount()-1)

	if z > 0:

		# determine shift
		deltaX = x - formerX
		deltaY = y - formerY

		# apply translation transformation
		shifted = ops.run("CLIJ_affineTransform", slice, "translatex=" + str(deltaX) + " translatey=" + str(deltaY))

		# copy result back
		ops.run("CLIJ_copySlice", inputGPU, shifted, z)

		ops.run("CLIJ_close", shifted)
	else:
		formerX = x
		formerY = y
	ops.run("CLIJ_close", slice)
	ops.run("CLIJ_close", binary)

ui.show("center of mass per slice", mass)

# show result
result = ops.run("CLIJ_pull", inputGPU)
ui.show("motion correction", result)

#cleanup
ops.run("CLIJ_close", inputGPU)
ops.run("CLIJ_close")

File: src/test/java/net/haesleinhuepf/clij/ops/examples/MotionCorrection.java

package net.haesleinhuepf.clij.ops.examples;

import java.io.IOException;

import org.junit.Test;
import org.scijava.table.DefaultGenericTable;
import org.scijava.table.GenericTable;

import net.haesleinhuepf.clij.ops.CLIJ_affineTransform.CLIJ_affineTransform;
import net.haesleinhuepf.clij.ops.CLIJ_centerOfMass.CLIJ_centerOfMass;
import net.haesleinhuepf.clij.ops.CLIJ_close.CLIJ_close;
import net.haesleinhuepf.clij.ops.CLIJ_copySlice.CLIJ_copySlice;
import net.haesleinhuepf.clij.ops.CLIJ_pull.CLIJ_pull;
import net.haesleinhuepf.clij.ops.CLIJ_push.CLIJ_push;
import net.haesleinhuepf.clij.ops.CLIJ_threshold.CLIJ_threshold;
import net.imagej.ImageJ;
import net.imglib2.img.Img;

public class MotionCorrection {

	@Test
	public void run() throws IOException {

		// define a threshold to differentiate object and background
		int threshold = 50;

		ImageJ ij = new ImageJ();

		Img input = (Img) ij.io().open(getClass().getResource(
			"/samples/motion_correction_Drosophila_DSmanila1.tif").getPath());
		ij.ui().show("input", input);

		Object input32 = ij.op().convert().float32(input);
		// push image to GPU
		Object inputGPU = ij.op().run(CLIJ_push.class, input32);

		double formerX = 0;
		double formerY = 0;

		GenericTable mass = new DefaultGenericTable();

		// process all slices; only the first stays where it is
		for (int z = 0; z < input.dimension(2); z++) {
			ij.log().info("z: " + z);

			Object slice = ij.op().run(CLIJ_copySlice.class, inputGPU, z);

			// determine center of mass
			Object binary = ij.op().run(CLIJ_threshold.class, slice, threshold);
			ij.op().run(CLIJ_centerOfMass.class, mass, binary);
			double x = (double) mass.get("MassX", mass.getRowCount() - 1);
			double y = (double) mass.get("MassY", mass.getRowCount() - 1);

			if (z > 0) {

				// determine shift
				double deltaX = x - formerX;
				double deltaY = y - formerY;

				// apply translation transformation
				Object shifted = ij.op().run(CLIJ_affineTransform.class, slice,
					"translatex=" + deltaX + " translatey=" + deltaY);

				// copy result back
				ij.op().run(CLIJ_copySlice.class, inputGPU, shifted, z);

				ij.op().run(CLIJ_close.class, shifted);
			}
			else {
				formerX = x;
				formerY = y;
			}

			ij.op().run(CLIJ_close.class, slice);
			ij.op().run(CLIJ_close.class, binary);
		}

		ij.ui().show("center of mass per slice", mass);

		// show result
		Object result = ij.op().run(CLIJ_pull.class, inputGPU);
		ij.ui().show("motion correction", result);

		// cleanup
		ij.op().run(CLIJ_close.class, inputGPU);


	}

	public static void main(String... args) throws IOException {
		MotionCorrection task = new MotionCorrection();
		task.run();
	}
}
